### DISCUSSION TITLE
Apache not found in providers folder

### DISCUSSION THREAD

**Discussion by erosons** [apache/airflow#27803] [General]
### Apache Airflow version

2.4.3

### What happened

After installing the providers package , I restarted my instance of airflow, I  was able to setup the spark connections  but  my dag sparksubmit operator cannot find the apache . spark folder path in the airflow provider path

Broken DAG: [/opt/airflow/dags/sparkSubmitPerMeter.py] Traceback (most recent call last):
  File "<frozen importlib._bootstrap>", line 228, in _call_with_frames_removed
  File "/opt/airflow/dags/sparkSubmitPerMeter.py", line 4, in <module>
    from airflow.providers.apache.spark.operators.spark_submit import SparkSubmitOperator
ModuleNotFoundError: No module named 'airflow.providers.apache'

Here is my code based

```
from airflow.models.dag import DAG
from datetime import datetime, timedelta
#from airflow.providers.apache.spark.operators.spark_sql import SparkSqlOperator
from airflow.providers.apache.spark.operators.spark_submit import SparkSubmitOperator
#from airflow.contrib.operators.spark_submit_operator import SparkSubmitOperator

default_args = {
    'owner': 'samson',
    'start_date': datetime(2022, 8, 20),
    'retries':2,
    'retry_delay': timedelta(hours=1),
}
with DAG('ETL_sparkSubmitOperatorV1',
                  default_args=default_args,
                  schedule_interval='@daily',
                  tags=["SparkSubmitTest"],
                  catchup=False

                ) as dag:

    # spark = SparkSession(SparkContext(conf=SparkConf()).getOrCreate())

    SparkSubmitOps= SparkSubmitOperator(
        task_id='sparkSubmitOps',
        conn_id='Spark-Conn',
        application="/opt/airflow/dags/amperon/sparkPerMeterLTF.py",
        total_executor_cores=3,
        executor_memory="2G",
        executor_cores=1,
        driver_memory="2G",
        num_executors=6,
        env_vars=None,
        dag=dag
    )

    SparkSubmitOps
~
```
~
~

### Operating System

Debian

### Versions of Apache Airflow Providers

apache-airflow-providers-apache-spark 4.0.0

### Deployment

Docker-Compose

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by boring-cyborg** [comment]
Thanks for opening your first issue here! Be sure to follow the issue template!

**Comment by o-nikolas** [comment]
Can you describe how you installed the spark provider? And if you still have it paste the output you got.

**Comment by erosons** [comment]
I did docker exec it <containerID> bash  into the airflow webserver
container and run the pip install apache-airflow-providers-apache-spark

Thank you
Samson

**Comment by Taragolis** [comment]
Due to information that you provided you only install additional packages into webserver container after started it however you need to install into all of containers/services (at least scheduler/worker).

This is not a bug of Airflow. I would suggest install packages during build image based on airflow image or customise airflow image.

Links below might help you:
- https://airflow.apache.org/docs/docker-stack/build.html#building-the-image
- https://airflow.apache.org/docs/docker-stack/build.html#customizing-the-image

**Reply by potiuk** [reply]
Yep. Installing packages by execing into them is not the right way if you work with containers. You need to build your own image https://airflow.apache.org/docs/docker-stack/build.html - it's a quite a bit of misunderstanding on how containers work if you try to install packages like you tried. You either do virtualenv approach and then you install packages as you did or - if you want go into container/docker world, you need to change the approach and understand that containers are immutable by default and you need to add new dependencies to thwm. This is an inherent property of containers and going into container world without grasping the container/image approach is likely a recipe for similar problems you have .

**Reply by erosons** [reply]
Thanks guys for your input so far , I try rebuilding the image and adding airflow provider packages to the requirements.txt.  the image got stuck and would not build and as soon as I remove the  airflow providers package, image got built successfully.

This got wondering if  airflow containerized version is designed to handle the provider packages?
Because , I experimented the whole process with a bare metal instance of airflow and everything seems to work  just fine.

**Reply by potiuk** [reply]
Yes it is. and Yes it works. If you follow examples from the docs - they are not only showing how to do it but also build automatically.

The problem you might have (I guess because `stuck` is not very specific description of the problem you haave) is that provider(s) you want to install are incompatible with other dependencies you already have. You have not written what `stuck` means but it might mean that `pip` is trying to find a good set of dependencies to match the packages you have and packages you install.

This is called `pip` backtracking and is described here https://pip.pypa.io/en/stable/topics/dependency-resolution/. This has nothing to do with airflow, however due to airlfow's complexity of dependencies (~ 700 dependencies in total for airflow and providers) this is not always possible to install specific version of airflow and specific versions of providers together - especially if you have other dependencies.

Unfortunately you need to solve your dependency conflicts. 

There are few ways that can help youL

1) use constraints when installing providers and other dependencies - https://airflow.apache.org/docs/apache-airflow/stable/installation/installing-from-pypi.html - constraints are set of "known good versions" of dependencies of Airlfow that we know works - we have separate constraints for different Airflow and Python combinations.  But ths will limit your providers and dependencies to specific versions

2) Enter the image `docker run -it <image> bash` and try to perform your `pip install` instructions from your Dockerfile manually - this will show you what `pip` does and where it backtracks and might help you to change or limit the dependencies you install to specific versions. Very sadly `pip`s output is often misleading and deoes not give you the right hint on how you can speed up the resolution and find good set of dependencies. Also the way how resolver algorithm works - you might need to add some extra hints `dependency>=2.0` in order to make the process of resolution faster (and sometimes in order to complete). Again - this is `pip` issue not `airflow`. And the reason is that dependency resolution is REALLY HARD in Python world. Unbelivably hard actually - so this is not `pip`s fault - they do their best to guess the right set of dependencies, but sometimes the algorithm needs  a bit of help.

3) you can also install your dependencies in smaller increments. Sometimes it is far harder job for `pip` to install 10 dependencies at once rather than split it into 10 separate steps to install each dependency sepearately

This is a bit of an art. The easiest way is to follow is 1) - there Airflow maintainers did all the heavy lifting and produced set of dependencies that are "cool".  Sometimes it means that you need to modify your code to use newer versions of dependencies - those that Airflow has in the constraints -  and this is often the fastest way to approach it.

**Reply by erosons** [reply]
Thank you Potiuk your feedback was really expository. I will look into the docs.
